metrix/src/app.rs

```diff
@@ -41,6 +41,89 @@ fn query_metric_route(
 ) -> Result<Json<Vec<Metric>>, BadRequest<String>> {
     let db_conn = establish_connection();
 
+    let filter_clause: String;
+    let result = build_filter_clause(offset, start_datetime, end_datetime, q);
+    match result {
+        Ok(o) => {
+            filter_clause = o;
+        },
+        Err(e) => {
+            return Err(e);
+        },
+    }
+
+    println!("### QUERY: SELECT * FROM metrics {};", filter_clause);
+
+    let query_string = format!("SELECT * FROM metrics {} ORDER BY id LIMIT 10", filter_clause);
+    let results = sql_query(query_string)
+        .load(&db_conn)
+        .expect("Error loading metrics");
+
+    Ok(Json(results))
+}
+
+#[get("/<aggregation>?<offset>&<start_datetime>&<end_datetime>&<q>&<bucket_count>")]
+fn aggregate_metrics_route(
+    aggregation: Option<&RawStr>,
+    offset: Option<&RawStr>,
+    start_datetime: Option<&RawStr>,
+    end_datetime: Option<&RawStr>,
+    q: Option<&RawStr>,
+    bucket_count: i32,
+) -> Result<Json<Buckets>, BadRequest<String>> {
+
+    let db_conn = establish_connection();
+
+    let filter_clause: String;
+    let result = build_filter_clause(offset, start_datetime, end_datetime, q);
+    match result {
+        Ok(o) => {
+            filter_clause = o;
+        },
+        Err(e) => {
+            return Err(e);
+        },
+    }
+
+    let start_timestamp= NaiveDateTime::parse_from_str(
+        &start_datetime.unwrap(),
+        &"%Y-%m-%dT%H:%M:%S".to_string()
+    ).ok().unwrap().timestamp();
+
+    let end_timestamp = NaiveDateTime::parse_from_str(
+        &end_datetime.unwrap(),
+        &"%Y-%m-%dT%H:%M:%S".to_string()
+    ).ok().unwrap().timestamp();
+
+    let bucket_size = (end_timestamp - start_timestamp) as f32 / bucket_count as f32;
+
+
+    let query_string = format!(
+        "SELECT
+            COUNT(*) as result,
+            FLOOR((extract(epoch from created_at)-{})/{})::INTEGER as bucket
```

Comment:
Explain this query pls

Reply:
Yeah so this line here is essentially what maps a given row to a bucket number between [0, bucket_count - 1].

So, first thing it does is convert `created_at` to an epoch so that we can work with the time as a proper integer. `extract(epoch from created_at)` does that.

Then it subtracts the start time given in the query as an epoch value to noramlize created_at within the range of 0 to (endtime - starttime) for bucketing. `extract(epoch from created_at) - {}` does that.

Then,  it actually gets bucketed. It does that by dividing the value above between [0, (endtime - starttime)] by the size of each bucket. This gives a floating point value between [0, bucket_count - 1]. Then, that value is FLOORed, which of course rounds down to the nearest integer. This is done so that every value gets pushed into a integer bucket. `FLOOR((extract(epoch from created_at)-{})/{}` does that.

Lastly, the query is filtered by whatever filters were specified in the request, and grouped by the bucket that it falls in. In the grouping, the count is applied, giving the result of the number of entries in each bucket (`as result`) and the bucket number (`as bucket`)